Tools/CISupport/ews-build/steps.py

```diff
@@ -7178,25 +7232,40 @@ def run(self):
         num_unexpected_results = self.getProperty('num_failing_files', 0) or self.getProperty('num_unexpected_issues', 0) or self.getProperty('num_passing_files', 0)
         unexpected_results_after_filter = None
         if num_unexpected_results:
-            logTextJson = self.log_observer_json.getStdout()
-            yield self._addToLog('stdio', f'Checking results database for unexpected results...\n')
-            successful_filter = yield self.filter_results_using_results_db(logTextJson)
-            unexpected_results_after_filter = self.getProperty('num_failing_files', 0) or self.getProperty('num_unexpected_issues', 0) or self.getProperty('num_passing_files', 0)
-            if successful_filter and self.was_filtered and unexpected_results_after_filter:  # If there are unexpected results
-                yield self._addToLog('stdio', f'\nSuccessfully filtered results! Updating unexpected_results.json on disk.\n')
-                self.write_unexpected_results_file_to_master()
-            if not successful_filter:
-                yield self._addToLog('stdio', f'\nFailed to consult results database. Falling back to tip-of-tree...\n')
-                # If results db failed, rebuild without changes to verify causation
-                self.build.addStepsAfterCurrentStep([ValidateChange(verifyBugClosed=False, addURLs=False), RevertAppliedChanges(exclude=['new*', 'scan-build-output*']), ScanBuildWithoutChange()])
-                self.result_message = self.createResultMessage()
-                return defer.returnValue(rc)
+            user_removed_tests = self.getProperty('user_removed_tests', [])
+            user_added_tests = self.getProperty('user_added_tests', [])
+
+            results_json = yield self.decode_results_data()
+            if not results_json:
+                return defer.returnValue(False)
+
+            unexpected_passes = self.get_unexpected_tests(results_json, 'passes') or ['Empty']
+            unexpected_failures = self.get_unexpected_tests(results_json, 'failures') or ['Empty']
+
+            if user_added_tests == unexpected_passes and user_removed_tests == unexpected_failures:
+                yield self._addToLog('stdio', 'Skipping second build since all unexpected results come from user modified expectations.\n')
+                unexpected_results_after_filter = True
```

Comment:
should `unexpected_results_after_filter` be True or False here? Also, can we do early return from this if statement?

Reply:
This should be True, since there are unexpected results (from line 7234, `if num_unexpected_results`), it just happens that they come from the user modified expectations and not from the checked in expectations.

I don't think we can exit early since there are various steps we want to append when there are unexpected results (around line 7269).